WordPress/gutenberg
# iAPI Router: Handle styles assets on region-based navigation

## What?

Adds support for style assets on region-based navigation, allowing stylesheets to be applied or removed when navigating to a new page with different styles.

## Why?

This feature is required for some core blocks like the Comments one. In this particular case, when publishing a new comment on a post that does not have any comments, we want to display the new comments without reloading the page using `navigate()`. The new comment might require styles not present on the initial post page.

## How?

The implementation uses the [`adoptedStyleSheets` API](https://developer.mozilla.org/en-US/docs/Web/API/Document/adoptedStyleSheets).

For each `<style>` and `<link rel=stylesheet>` tags, a `CSSStyleSheet` is generated. If the CSS code contains a relative URL pointing to another asset (e.g., images or fonts), the URL is converted into an absolute URL. That could be removed once the `baseURL` option is supported in Chrome, Edge, or Opera (see [coverage](https://developer.mozilla.org/en-US/docs/Web/API/CSSStyleSheet#:~:text=options.baseURL%20parameter) in MDN).

These `CSSStyleSheet` instances are then passed to the `adoptedStyleSheets` property in the correct order.

With these APIs, we can solve the following problems:
- Wait for all the styles to be ready before updating the HTML.
- Update both the styles and the HTML at the same time.
- Reuse all previous stylesheets without having to parse the CSS code again.

The PR also replaces the current implementation we have for full-site navigation.

## Testing Instructions

1. In the site editor, go to the home template
2. Ensure the "Force page reload" setting is disabled in the Query block.
3. Add an image block inside the Post Template.
4. Change its style and make the image rounded.
5. Visit a page of the home that doesn't exist (e.g., page 2) so it shows the "No Results" block.
6. Click on the Previous Page link.
7. Images should appear rounded.

**Important:** you could notice elements outside the Query block that their appearance change. It's related to [this issue](https://github.com/WordPress/gutenberg/issues/67308), and will be addressed in a subsequent PR.
